deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/ocnn/OCNNOutputLayer.java

```diff
@@ -0,0 +1,110 @@
+package org.deeplearning4j.nn.conf.ocnn;
+
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.NoArgsConstructor;
+import lombok.ToString;
+import org.deeplearning4j.nn.api.Layer;
+import org.deeplearning4j.nn.api.ParamInitializer;
+import org.deeplearning4j.nn.conf.NeuralNetConfiguration;
+import org.deeplearning4j.nn.conf.layers.BaseOutputLayer;
+import org.deeplearning4j.nn.conf.layers.LayerValidation;
+import org.deeplearning4j.nn.layers.ocnn.OCNNParamInitializer;
+import org.deeplearning4j.optimize.api.TrainingListener;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.shade.jackson.annotation.JsonCreator;
+import org.nd4j.shade.jackson.annotation.JsonIgnoreProperties;
+import org.nd4j.shade.jackson.annotation.JsonProperty;
+
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * An implementation of one class neural networks from:
+ * https://arxiv.org/pdf/1707.02131.pdf
+ *
+ * The one class neural network approach is an extension of the standard output layer
+ * with a single set of weights, an activation function, and a bias to:
+ * 2 sets of weights, a learnable "r" parameter that is held static
+ * 1 traditional set of weights.
+ * 1 additional weight matrix
+ *
+ * @author Adam Gibson
+ */
+@Data
+@NoArgsConstructor
+@ToString(callSuper = true)
+@EqualsAndHashCode(callSuper = true)
+@JsonIgnoreProperties("lossFn")
+public class OCNNOutputLayer extends BaseOutputLayer {
+    //embedded hidden layer size
+    //aka "K"
+    private int hiddenSize;
+
+    private double nu = 0.04;
+
+    public OCNNOutputLayer(Builder builder) {
+        super(builder);
+        this.hiddenSize = builder.hiddenLayerSize;
+        this.nu = builder.nu;
+
+    }
+
+    @JsonCreator
+    public OCNNOutputLayer(@JsonProperty("hiddenSize") int hiddenSize,@JsonProperty("nu") double nu) {
+        this.hiddenSize = hiddenSize;
+        this.nu = nu;
+    }
+
+    @Override
+    public Layer instantiate(NeuralNetConfiguration conf, Collection<TrainingListener> trainingListeners, int layerIndex, INDArray layerParamsView, boolean initializeParams) {
+        LayerValidation.assertNInNOutSet("OCNNOutputLayer", getLayerName(), layerIndex, getNIn(), getNOut());
+
+        Layer ret = new org.deeplearning4j.nn.layers.ocnn.OCNNOutputLayer(conf);
+        ret.setListeners(trainingListeners);
+        ret.setIndex(layerIndex);
+        ret.setParamsViewArray(layerParamsView);
+        Map<String, INDArray> paramTable = initializer().init(conf, layerParamsView, initializeParams);
+        ret.setParamTable(paramTable);
+        ret.setConf(conf);
+        return ret;
+    }
+
+    @Override
+    public ParamInitializer initializer() {
+        return OCNNParamInitializer.getInstance();
+    }
+
+
+    @Override
+    public double getL1ByParam(String paramName) {
+        return l1;
```

Comment:
r should be excluded from L1 and L2, I would think?

Reply:
@AlexDBlack  I don't think so here..I would like the ability to add l2/l1 on top of it. The R is actually part of SGD as a learned parameter.